customlists/customlist_explain.py

```diff
@@ -0,0 +1,228 @@
+import argparse
+import csv
+import json
+import logging
+from typing import List
+
+from customlists.customlist_report import (
+    CustomListProblemBookBrokenOnSourceCM,
+    CustomListProblemBookMismatch,
+    CustomListProblemBookMissing,
+    CustomListProblemBookRequestFailed,
+    CustomListProblemCollectionMissing,
+    CustomListProblemCollectionRequestFailed,
+    CustomListProblemListAlreadyExists,
+    CustomListProblemListBroken,
+    CustomListProblemListUpdateFailed,
+    CustomListReport,
+    CustomListsReport,
+)
+
+
+class CustomListImportExplainer:
+    @staticmethod
+    def _parse_arguments(args: List[str]) -> argparse.Namespace:
+        parser: argparse.ArgumentParser = argparse.ArgumentParser(
+            description="Explain what went wrong during an import."
+        )
+        parser.add_argument(
+            "--verbose",
+            "-v",
+            action="count",
+            default=0,
+            help="Increase verbosity",
+        )
+        parser.add_argument(
+            "--report-schema-file",
+            help="The schema file for custom list reports",
+            required=False,
+            default="customlists/customlists-report.schema.json",
+        )
+        parser.add_argument(
+            "--report-file",
+            help="The report file that was produced during importing",
+            required=True,
+        )
+        parser.add_argument(
+            "--output-csv-file",
+            help="The output CSV file containing the list of books to be fixed",
+            required=True,
+        )
+        return parser.parse_args(args)
+
+    def _load_report(self) -> CustomListsReport:
+        _schema_dict = {}
+        with open(self._report_schema_file, "rb") as f:
+            _schema_dict = json.load(f)
+        _report_dict = {}
+        with open(self._report_file, "rb") as f:
+            _report_dict = json.load(f)
+        return CustomListsReport.parse(_schema_dict, _report_dict)
+
+    def _generate_csv_for_list(self, csv_writer, list_report: CustomListReport) -> int:
+        wrote_rows = 0
+        for problem in list_report.problems():
+            ++wrote_rows
+            if isinstance(problem, CustomListProblemBookBrokenOnSourceCM):
+                csv_writer.writerow(
+                    [
+                        list_report.name(),
+                        problem.id(),
+                        problem.title(),
+                        problem.author(),
+                        type(problem).TYPE,
+                        problem.message(),
+                        "The book could not be retrieved on the source CM, but it might be available with a manual search "
+                        "on the target CM.".replace("\n", ""),
+                    ]
+                )
+                continue
```

Comment:
Do we need `continues` everywhere or can this be if-elif-else(raise). While adding new problem cases people are going to forget to add continues, which would be an annoying error to debug.

Reply:
It could be `if-elif-raise`. I'll make that change now.

Reply:
How I wish we had algebraic data types and pattern matching... :weary: